packages/fcl/src/wallet-provider-spec/user-signature.md

```diff
@@ -0,0 +1,266 @@
+- **Improved Application Login**
+  - **Increased security**: Arguably more secure than proof of ownership by email/password
+  - **Simplified UX**: No application password required
+  - **Increased privacy**: No email or third party authentication service needed
+- **Message Validation**: Assuring that a message sent or received has not been tampered with
+- **Multisig contracts**
+- **Decentralised exchanges**
+- **Meta transactions**
+
+
+# Config and Authentication
+
+As a prerequisite, **FCL** is configured to point to the Wallet Provider's Authentication Endpoint.
+
+> During development (and on mainnet) FCL can be configured to use the wallet directly by
+> setting the **Challenge Handshake Url** to the wallet providers **Authentication Endpoint**
+> by configuring fcl like this `config().put("challenge.handshake", "https://my-awesome-wallet-provider.com/fcl/authenticate")`.
+
+1. A user initiates authentication with the wallet provider via application UI
+2. The wallet confirms a user's identity and sends back information used to configure **FCL** for future user actions in the application
+3. Included in the authentication response should be the provider's [Key Services](#) including a **`user-signature`** service for use with **`signUserMessage()`**
+
+
+# User Signature Service
+
+A User Signing Service's job is to receive a signable message from **FCL**, and send back a **`Composite Signature`**.
+
+A **`user-signature`** service should conform to the following:
+
+```javascript
+{
+  f_type: "Service",
+  f_vsn: "1.0.0",
+  type: "user-signature",
+  uid: "uniqueDedupeKey",
+  endpoint: "http://rawr.io",
+  method: "IFRAME/RPC",
+  uid: "____",
+  endpoint: "https://____",
+  method: "IFRAME/RPC", 
+  "id": "____", // wallets internal id for the user
+  data: {},   // included in body of user-signature request
+  params: {}, // included as query params on endpoint url
+}
+```
+
+The `user-signature` service returns a **`PollingResponse`** with a **`CompositeSignature`** or `null` as the data.
+
+
+## `PollingResponse`
+
+```javascript
+{
+  f_type: "PollingResponse",
+  f_vsn: "1.0.0",
+  status: "APPROVED", // PENDING | APPROVED | DECLINED
+  reason: null,      // if status is DECLINED this is a string that specifies why
+  data: null,        // if status is APPROVED this is the value FCL needs
+  updates: null,     // Optional `back-channel-rpc` Service (Required if status is PENDING)
+  local: null,       // Optional `frame` Service
+}
+```
+
+## `CompositeSignature`
+
+```javascript
+{
+  f_type: "CompositeSignature",
+  f_vsn: "1.0.0",
+  addr: "____", // Flow Address (sans-prefix)
+  keyId: 3,
+  signature: "______", // Signature as a hex string
+}
+```
+
+# Signing Sequence
+
+1. Application sends message to signing service. **FCL expects a hexadecimal string**
+3. Service prepares, hashes, and signs the message
+2. Wallet returns Composite Signature consisting of **addr**, **keyId**, **signature**
+
+---
```

Comment:
I think we can simplify this. If we need to explain all the different stuff every time we are in trouble.

Service configures fcl.
Fcl uses the services method to decide how to send the signable to the wallet
Wallet tags the signable.message and hashes it, signs it
Wallet sends back (makes available) a composite signature of the tagged and hashed value.